File: services/python/fastapi/src/routes/handlers/post.py
import structlog
from uuid import uuid4, UUID
from fastapi import Request
from typing import TypedDict

from app.task import Task
from app.state import AppState


class TaskId(TypedDict):
    id: str


async def post_handler(request: Request, task: Task) -> TaskId:
    logger = structlog.get_logger()

    id: UUID = uuid4()
    state: AppState = request.app.state.app_state

    async with request.app.state.tasks_lock:
        state.tasks[id] = task

    logger.info(
        "Inserted new task",
        id=str(id),
        secret=len(task.secret),
        operation=task.operation.lower(),
        method="POST",
    )

    return {"id": str(id)}

File: services/go/gin/routes/handlers/post.go
package handlers

import (
	"app"
	"log/slog"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

func PostHandler(state *app.AppState) gin.HandlerFunc {
	return func(c *gin.Context) {
		body, err := ReadBody(c)
		if err != nil {
			return
		}

		task, err := ParseTask(c, body)
		if err != nil {
			return
		}

		id := uuid.New()

		state.Mu.Lock()
		state.Tasks[id] = task
		state.Mu.Unlock()

		slog.Info(
			"Inserted new task",
			"id", id.String(),
			"secret", len(task.Secret),
			"operation", strings.ToLower(string(task.Operation)),
			"method", c.Request.Method,
		)

		c.JSON(http.StatusCreated, gin.H{"id": id.String()})
	}
}
